cpp/src/arrow/extension/fixed_shape_tensor.cc

```diff
@@ -0,0 +1,267 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/extension/fixed_shape_tensor.h"
+
+#include "arrow/array/array_nested.h"
+#include "arrow/array/array_primitive.h"
+#include "arrow/json/rapidjson_defs.h"  // IWYU pragma: keep
+#include "arrow/tensor.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/sort.h"
+
+#include <rapidjson/document.h>
+#include <rapidjson/writer.h>
+
+namespace rj = arrow::rapidjson;
+
+namespace arrow {
+namespace extension {
+
+bool FixedShapeTensorType::ExtensionEquals(const ExtensionType& other) const {
+  if (extension_name() != other.extension_name()) {
+    return false;
+  }
+  const auto& other_ext = static_cast<const FixedShapeTensorType&>(other);
+  bool equals = storage_type()->Equals(other_ext.storage_type());
+  equals &= shape_ == other_ext.shape();
+  equals &= permutation_ == other_ext.permutation();
+  equals &= dim_names_ == other_ext.dim_names();
```

Comment:
Maybe not that important, but if you do this in one expression, it will shortcut?

```
return storage_type()->Equals(other_ext.storage_type()) && shape_ == other_ext.shape() && permutation_ == other_ext.permutation() && dim_names_ == other_ext.dim_names();
```

Reply:
I suppose it would. Will change and also add a check where `permutation.empty() == permuation([0, 1, 2])`.